Implementation of the data model for v3.8

jeff5 opened the pull request:
This is a proposed implementation for objects in the core of Jython 3. It is the model I talked about at the on-line [Python Language Summit 2021](https://pyfound.blogspot.com/2021/05/the-2021-python-language-summit_22.html), applied to the single Python type `int`. It is foundational for, but does not include, the use of dynamic call sites for performance.

It is draft as I'd like to push more onto it before it gets into the main repo. I thought it good to have it out there as a PR for those that want to follow along.

It is somewhat unsatisfactory as a PR because it includes a lot of code that has no purpose without parts I intend to leave to later PRs. I'll add more tests, I think, and maybe the exposure mechanism, but avoid the trap of piling everything in one massive PR.

jeff5 committed:
Foundation classes of the v3.8 data model

This change begins a shift to a "plain Java objects" approach to the
data model. We provide PyType and Operations classes, the marker
interfaces and a base Python object implementation. The elaborate Slots
enum describes a system of handles for special functions used by the
run-time. Other classes, although numerous, are present to support these
and errors they raise.

There is a unit test that exercises creation and lookup of each broad
sub-class of Operations object.

jeff5 committed:
[Build fails] Move PyLong and parts of stringlib from 2 to 3

Move these Jython 2 source files unmodified to the Jython 3 source tree,
preparatory to conversion for Jython 3. (Accept the build fails.)

As a rule, each committed state should build without error. As an
exception, we commit pure file moves from 2 to 3 to provide an
unambiguous pedigree to git (and code archaeologists).

jeff5 committed:
Get the re-used Jython 2 code to compile

The changes to convert PyLong from 2 to 3 are large. Formatting code in
stringlib can be re-used with relatively minor changes. This change set
also brings in new classes for exceptions, and stubs for certain core
classes to satisfy dependencies.

We preserve legacy annotations in PyLong as comments so that the git
diff can align the old and new implementations of corresponding methods.

jeff5 committed:
Support Python in the Gradle build

We will use Python to generate some of our code, and it has to be a 3.8.
We adjust the Travis CI build to support that, and insert a gratuitous
use of Python as a test.

jeff5 committed:
Add the bool type and extend a test to it

bool provides a simple case of inheritance but it is rather a special
one since Java Boolean is its canonical implementation. We extend the
Operations formation test to bool and int.

jeff5 committed:
Formatting change to PyType

Allow code to be wide, but keep comments at a readable width. This aids
comparison with the VSJ original, while we need it.

jeff5 committed:
Expose special functions as slots in Operations

We provide an exposure mechanism sufficient to populate the dictionary
of the type object for any built-in, and from that the slots of
Operations objects. We apply this to int and bool.

We test that we can invoke a few unary special methods for int and bool.

jeff5 committed:
Add a test of binary slot wrappers and refactor

The refactoring extracts common code between the unary and binary tests
(and later ones to be developed). The testing is at the level of
descriptors: we are not yet caching these as method handles in the
Operations object, although much of the code is present.

jeff5 committed:
Begin to fill and use the op_* slots

This change adds the code to Slot.java that causes the slots of
Operations objects to be filled with handles to implementation methods.
The unary and binary slot wrapper tests are extended to invoke these
slots and test the result against the same criteria as for the
corresponding descriptor.

jeff5 committed:
Begin to generate method specialisations in PyLongMethods

This change imports Java code generation in Python and applies it to the
repetitive task of creating a version of each PyLong method specialised
for the class 'self'. Implementations inherited from Jython 2, still
present for now, are ignored where a class-specific duplicate exists.

Stewori commented on core/src/main/javaTemplate/org/python/core/PyLongMethods.java:
```diff
@@ -0,0 +1,81 @@
+     */
+    static Object toInt(BigInteger r) {
+        /*
+         * Implementation note: r.intValueExact() is for exactly this
```
Maybe a use case for [@implNote](https://openjdk.java.net/jeps/8068562). Despite the draft status the described tags are already in frequent use in current and recent Java standard API doc since several years.

jeff5 replied:
You're right and I've used it elsewhere in the code. It needs some [special pleading in the javadoc command](https://github.com/jeff5/jython/blob/8fcaec5b0d3ce4bd7bf39d6cc77e9a6f57f5b515/core/core.gradle#L99-L108) to enable it. Probably, this text is from before I learned that.

jeff5 replied:
Actually, in context, I see this is regular comment, not Javadoc. It is definitely a note to maintainers, not to implementers of an API, explaining why I did not use the provided method. There are probably bits of Javadoc that could do with `@implNote` but none so obvious. The JEP is a good reminder how to use the tags.

Stewori commented on core/src/main/java/org/python/core/PyLong.java:
```diff
@@ -0,0 +1,1415 @@
+    /*
+
+    @ExposedGet(name = "real", doc = BuiltinDocs.long_real_doc)
+     */
```
It occurred to me that dual docs for Java and Python docstrings could be handled using Javadoc's [@value](https://docs.oracle.com/en/java/javase/16/docs/specs/javadoc/doc-comment-spec.html#value) tag.
E.g. in this case the javadoc would be
```
/**
 * {@value BuiltinDocs#long_real_doc}
 */
```

jeff5 replied:
That's a clever idea. The actual doc string in code is not quite what you see in help (has some mark-up) but is close enough to be readable by maintainers.

These annotations from Jython 2 are preserved in this odd form to make it easier to generate a diff that aligns with the old code. (`kdiff3` needed these clues.)

I have equivalent annotations in the experimental code off-repo, including one for docstrings, but I'm not sure how it should work. Something similar to Jython 2 would be good, with the objective that strings don't load until used. I couldn't tell if Jython 2 achieves that. It's not my top thing to work on.

jeff5 committed:
Implement int division and modulus

These are evolved versions of the Jython 2 code, moved to PyLongMethods,
and called from generated wrappers. These operations are not simply
invocations of the Java counterparts but quite complex, so we add a
specific test.

jeff5 committed:
Add all the special method slots and Javadoc them

Slots are added to Slot and OPerations based on CPython slots and how
Jython works. Many of these are unused, but where they match methods in
the ported PyLong, the signature will be validated. Since documentation
is a focus, this change also fixes the javadoc so it builds cleanly.

jeff5 committed:
Rely on generated versions of the special methods (slots)

Many of the special methods were already in the generated PyLongMethods,
but now we remove them absolutely from PyLong. We add tests that call a
few of them, to prove this works.

We implement __(r)pow__ and the __*shift__ methods as hand-crafted
additions to PyLongMethods, rather than generating them, as they are
irregular.

jeff5 committed:
Copyright notices and some formatting

We reproduce on the new code the style of copyright notice found
elsewhere in Jython, with the addition of "Licensed to PSF ...",
requested in the  contributor agreement we all signed.

The copyright label is probably not necessary, but we follow precedent.

jeff5 commented:
I think this is ready for inspection now.

Obviously it is nowhere near a Python interpreter, but only illustrates an aspect of the proposed new implementation, namely how we would invoke special methods (slots) on built-in types. This, however, is fundamental.

And not that easy to follow, I should think, so here is a little code-walk of the idea.

The best way to appreciate what's going on may be to run in your IDE the unit test [AbstractAPITest](https://github.com/jeff5/jython/blob/8952c3e9afb375357157257ad48bcd1035fdd02b/core/src/test/java/org/python/core/AbstractAPITest.java#L27), then single step one of its cases under the debugger, let's say the test of `Abstract.repr`. Now, you should reach a breakpoint you set in [PyLong.__repr__](https://github.com/jeff5/jython/blob/8952c3e9afb375357157257ad48bcd1035fdd02b/core/src/main/java/org/python/core/PyLong.java#L440-L444). You can maybe see how that got called.

For a more complicated case, consider debugging the test of `PyNumber.negative` in [AbstractNumberAPITest](https://github.com/jeff5/jython/blob/8952c3e9afb375357157257ad48bcd1035fdd02b/core/src/test/java/org/python/core/AbstractNumberAPITest.java#L28) that tests `__neg__`. You would have to set breakpoints in one of the implementations of `PyLongMethods.__neg__` in the _generated_ code. (There is no source for that until you run the build.) How does Jython select the signature matching either `Integer` or `BigInteger` arguments?

The action depends on Jython finding `MethodHandle`s for the implementation methods in structures equivalent to the CPython type object, so before that can work, we must populate those slots from the object implementations. This code finds the implementations:
https://github.com/jeff5/jython/blob/8952c3e9afb375357157257ad48bcd1035fdd02b/core/src/main/java/org/python/core/TypeExposer.java#L84-L95
and this code creates descriptors in the dictionary of the type (here `int`):
https://github.com/jeff5/jython/blob/8952c3e9afb375357157257ad48bcd1035fdd02b/core/src/main/java/org/python/core/TypeExposer.java#L66-L74.
Back in the type object itself, the slots are filled here:
https://github.com/jeff5/jython/blob/8952c3e9afb375357157257ad48bcd1035fdd02b/core/src/main/java/org/python/core/PyType.java#L586-L591

thetric commented on build.gradle:
```diff
@@ -39,15 +39,22 @@ group = 'org.python'
     // Same configuiration all sub-projects
     tasks.withType(JavaCompile) {
         options.encoding = 'UTF-8'
         options.debug = true
```
Perhaps you want to set `--release 11` too?

> --release release
> 
> Compiles against the public, supported and documented API for a specific VM version. Supported release targets are 6, 7, 8, 9, 10, and 11.
> 
> Note:
> 
> When using --release for a version of the Java Platform that supports modules, you can’t use --add-modules to access internal JDK modules, nor can you use --add-exports to access internal JDK APIs in the modules.
>
> https://docs.oracle.com/en/java/javase/11/tools/javac.html#GUID-AEEC9F07-CB49-4E96-8BC7-BCC2C7F725C9

jeff5 replied:
That seems a useful discipline to accept. The magic is:
```
    tasks.withType(JavaCompile) {
        // Use only public API
        options.compilerArgs.addAll(['--release', '11'])
        ...
    }
```
I'll slip it into `main` later.

jeff5 commented:
Everyone seems content with this, so I'll rebase-merge, in this case keeping the commits as distinct, rather than squash. I expect this to "fast-forward", but the big green button always feels like a leap into the unknown.

(I intend to take @thetric's suggestion as a separate change.)

jeff5 commented:
Process note: GitHub has not created a needless merge commit, but it did needlessly reproduce all the commits as new objects. :( Harmless in the long run, but I believe I could have avoided this at the prompt.
